LG-12858 - Add new vtr option (enhanced ipp)

gina-yamada opened the pull request:
## 🎫 Ticket

[LG-12858](https://cm-jira.usa.gov/browse/LG-12858)  Implementation: Add new Vector of Trust Option on Sinatra for F&F testing

## 🛠 Summary of changes
+ Set eipp_allowed in staging in oidc-sinatra*
+ Set eipp_allowed to false in all other environments in oidc-sinatra*
        * I think this is achieved. Please see changes in `.env.example` and `config.rb`
+ Added `Enhanced In-Person Proofing` option to the ial (level of service) drop down conditionally (to display in staging only)
+ Mapped new ial (enhanced-ipp-required) to vtr_values P1.Pe  (`P1` : Identity proofing is performed `Pe` :Enhanced proofing)
+ Mapped new ial to arc_value

## 📜 Testing Plan
1. Pull down branch `keithw/LG-13235-consume-eipp-data-from-sinatra` from idp. Open file `app/services/service_provider_request_handler.rb` and add `puts sp_session` on the first line in decorated_sp_session.
2. Start server on branch `keithw/LG-13235-consume-eipp-data-from-sinatra`
3. Pull down this branch locally.
4. Open .env and ensure `eipp_allowed` is set to false
5. Start the server.
6. Visit `http://localhost:9292/`
7. Click on the Level of Service drop down. You should not see `Enhanced In-Person Proofing` as an option.
8. Kill the server.  Change the value of `eipp_allowed` to true. Start the server.
9. Click on the Level of Service drop down. You should see `Enhanced In-Person Proofing` as an option now.
10. Change the Level of Service drop down to `Enhanced In-Person Proofing` and click sign in. 
11. You should be taken to Login locally
12. Inspect the identity-oidc-sinatra terminal. You should see the vtr (vector of trust contains value `C1.P1.Pe` (Pe= enhanced proofing) and that the ial is set to `enhanced-ipp-required`
13. Inspect the identity-idp terminal for the log you set up in testing plan 1. You should see that vrt is added to the session (that is Keith's PR) and that ["C1.P1.Pe"] is set (look at last line of log).
14. Visit `http://localhost:9292/` again.
15. Confirm we are not changing existing behavior. Pick other options in the Level of Service drop down.  Ensure that you are not seeing  `C1.P1.Pe` value be passed in when `Enhanced In-Person Proofing`  is not selected (ie you should see Pb for biometrics rather than Pe) and that the ial set is not `enhanced-ipp-required`.  (Step 12 and 13)

## 📷 Screenshots of env variable set in Cloud.gov [env]-identity-oidc-sinatra
<img width="497" alt="Screenshot 2024-05-31 at 9 53 07 AM" src="https://github.com/18F/identity-oidc-sinatra/assets/125507397/b6a1cc29-2f6d-4e79-82a6-a4eba6b55449">

<img width="502" alt="Screenshot 2024-05-31 at 9 52 51 AM" src="https://github.com/18F/identity-oidc-sinatra/assets/125507397/fe3c066b-8b9a-4abe-8c51-ae99396435ff">

<img width="589" alt="Screenshot 2024-05-31 at 9 51 45 AM" src="https://github.com/18F/identity-oidc-sinatra/assets/125507397/f1d82ca6-a153-4f2a-ba4a-1d435275d26e">

<img width="526" alt="Screenshot 2024-05-31 at 9 51 17 AM" src="https://github.com/18F/identity-oidc-sinatra/assets/125507397/a7bf094d-99c3-4818-926a-5cade1964784">

zachmargolis commented on app.rb:
```diff
@@ -49,6 +49,9 @@ def config
 
       ial = prepare_step_up_flow(session: session, ial: ial, aal: aal)
 
+      is_eipp_allowed = is_eipp_allowed()
+      ial_select_options = get_ial_select_options(is_eipp_allowed: is_eipp_allowed)
```
1. The empty parens for the method call jumps out at me, would definitely avoid that
2. See other comments, but I think we don't need to pass in `is_eipp_allowed`
3. I bet we could remove this entirely and just call the method inline in the hash below
```suggestion
```

gina-yamada replied:
I removed the method entirely and called the method inline in the hash with commit db8c6741ec7d24e4524248238bab00ae45f1fbc4.
That reads nicer.

zachmargolis commented on app.rb:
```diff
@@ -145,6 +149,28 @@ def config
+      end      
+    end 
+
+    def is_eipp_allowed()
+      env = ENV['idp_environment']
+      is_eipp_allowed = (env == 'dev' || env == 'staging')
+    end
```
1. I'd drop the `is_` and name this with a `?` to match the previous comment
2. No need for the final assignment, the result is just the value
```suggestion
    def eipp_allowed?
      env = ENV['idp_environment']
      env == 'dev' || env == 'staging'
    end
```

zachmargolis replied:
Another thought: hardcoding specific environment names in source code is a [12-factor app antipattern](https://12factor.net/config#:~:text=In%20a%20twelve,over%20its%20lifetime.), could we add a new config for `ENV['eipp_allowed']` and set that in the appropriate environments?

gina-yamada replied:
Thank you for the recommendation. I think that is a better solution. Resolved with commit db8c6741ec7d24e4524248238bab00ae45f1fbc4.

zachmargolis commented on app.rb:
```diff
@@ -145,6 +149,28 @@ def config
+      else
+        option
+      end      
+    end 
```
1. since this is all in the same class, this method has access to the helper method for eipp so we could probably remove the param entirely (and I'd also inline the options just to keep things scoped
2. I'd pluralize to `options` since it's an array

```suggestion
    def ial_select_options
      option = [
        ['1', 'Authentication only (default)'],
        ['2', 'Identity-verified'],
        ['0', 'IALMax'],
        ['step-up', 'Step-up Flow'],
        ['biometric-comparison-required', 'Biometric Comparison (Disabled in prod)'],
      ]

      if eipp_allowed?
        options << ['enhanced-ipp-required', 'Enhanced In-Person Proofing (Enabled in dev & staging only)']
      else

      options
    end 
```

gina-yamada replied:
Resolved with commit ae25babbbada2aa7084ea3143a788d20e7e7e9c0

JackRyan1989 commented on app.rb:
```diff
@@ -193,6 +222,8 @@ def scopes_for(ial)
         'openid email x509'
       when '2', 'biometric-comparison-required'
         'openid email profile social_security_number phone address x509'
+      when '2', 'enhanced-ipp-required'
```
This duplicate when clause is going to be ignored. Is the ial '2' for enhanced-ipp?

gina-yamada replied:
Done. Resolved with commit c23a68dc35af6686a7f18a66e8062544a24f5ec0

eileen-nava reviewed (approved):
I tested this locally with Keith's branch from `identity-idp` and it worked as expected.  LGTM!

(Caveat: I'm still getting back into the swing of things on Joy, which means it might be helpful to have another Joy developer review this, too.)

gina-yamada commented:
> I tested this locally with Keith's branch from `identity-idp` and it worked as expected. LGTM!
> 
> (Caveat: I'm still getting back into the swing of things on Joy, which means it might be helpful to have another Joy developer review this, too.)

Thanks for the review! I prefer 2 when it is somethi

gina-yamada commented:
@KeithNava @eileen-nava  I am re-requesting a review because I made what I think are fairly large changes to the conditional logic of the drop down.  Pay extra attention to the last commit (see default_config func in config.rb). I think this is how we should go about this but not 100% confident.  Please review and fire up again. Thanks!

zachmargolis commented on config.rb:
```diff
@@ -77,13 +81,18 @@ def default_config
         env = ENV['idp_environment'] || 'int'
         domain = ENV['idp_domain'] || 'identitysandbox.gov'
 
+        if env == 'staging'
+          data['eipp_allowed'] = true
+        end
```
IMO ideally we would rely on the `eipp_allowed` config being correct in staging, rather than checking against the env here

JackRyan1989 replied:
@zachmargolis: @gina-yamada and I struggled with figuring out how to actually set the configuration value in staging. Could we maybe do a quick chat to see where it's set?

gina-yamada replied:
I am not sure either. I am poking around.

JackRyan1989 replied:
Ahhh cloud.gov. Got it. I wonder if we have access to the cloud.gov dashboard for this site? That should come along with github repo access, right?

zachmargolis replied:
We may have to invite you to that separately, we can follow up in Slack

gina-yamada replied:
I recently got access. I am looking around for the process. It looks like I can just add the new variable in staging and then  maybe/likely "restaging the app".  Looking into that.  (It recreates your app using the most recent environment.) I will update the code once this is done.

zachmargolis commented on app.rb:
```diff
@@ -250,6 +272,11 @@ def biometric_comparison_required_value(ial)
       ial == 'biometric-comparison-required'
     end
 
+    def enhanced_ipp_required_value(ial)
```
this is returning a boolean so let's make it a `?` method, and maybe even rename it to clearer about what is being checked?
```suggestion
    def requires_enhanced_ipp?(ial)
```

gina-yamada replied:
resolved with commit 6c529a1fc3f76e9897765cc2b0d92928a0c48e74

zachmargolis reviewed (approved):
LGTM with some suggestions

eileen-nava commented on app.rb:
```diff
@@ -147,18 +147,17 @@ def config
-      
-      if eipp_allowed == 'true'
+
+      if config.eipp_allowed?
```
👏🏻 I like this refactor.

eileen-nava commented on app.rb:
```diff
@@ -250,6 +272,11 @@ def biometric_comparison_required_value(ial)
       ial == 'biometric-comparison-required'
     end
 
+    def requires_enhanced_ipp?(ial)
```
I’d expect this method to return `nil` when it returns from line 276.  I know there's [a Ruby style convention ](https://rubystyle.guide/#bool-methods-qmark) that if a method returns a boolean, then the method should end with `?`.     Should a method also end with `?` if it returns a boolean or `nil`?

gina-yamada replied:
I am now returning false, see commit d9607e4706256f48603e3703c5cf26b5559b0e71

eileen-nava reviewed (approved):
I tested this locally and it worked well.  Approved.

zachmargolis commented on app.rb:
```diff
@@ -257,6 +279,11 @@ def biometric_comparison_required_value(ial)
     end
 
+    def requires_enhanced_ipp?(ial)
+      return unless config.vtr_disabled?
```
maybe we add an explicit `false` here to address [Eileen's comment?](https://github.com/18F/identity-oidc-sinatra/pull/161/files#r1612006986)
```suggestion
      return false unless config.vtr_disabled?
```

gina-yamada replied:
Done. Resolved with commit d9607e4706256f48603e3703c5cf26b5559b0e71

gina-yamada commented:
@eileen-nava @KeithNava  This is ready for a final review! Sorry for the delay. 

I updated the solution to use env variable set in Cloud.gov as suggested by @zachmargolis. 🙏 💯    I added screenshots above for each env so you can see how "eipp_allowed" is set in each environment.  I also updated the testing instructions to fit this new approach.  

I request review again primary to ensure how I am fetching that env variable is accurate. It should default to false (even though I set in all envs) and I am expecting it to come in as a string. I was using `vtr_disabled` as an example.  

Thanks again.
